csp_billing_adapter_microsoft/plugin.py

```diff
@@ -142,18 +191,86 @@ def _fetch_metadata(url):
         method='GET'
     )
     try:
-        value = urllib.request.urlopen(data_request).read()
-        return value.decode()
+        with urllib.request.urlopen(data_request) as value:
+            return value.read().decode("utf-8")
     except urllib.error.URLError as error:
-        log.error('Failed to retrieve metadata for: {url}. {error}'.format(
-            url=url,
-            error=str(error)
-        ))
-        return {}
-
-
-def _get_api_token():
-    """
-    Get the token to authenticate when using the Billing API
-   """
-    return json.loads(_fetch_metadata(TOKEN_URL))
+        log.error('Failed to retrieve metadata for: %s: %s', url, str(error))
+        return "{}"
+
+
+def _get_msi_token():
+    """Get the MSI token to authenticate when using the Billing API"""
+    # https://learn.microsoft.com/en-us/partner-center/marketplace/marketplace-metering-service-authentication
+
+    # Set resource id to the required value needed to to retrieve an
+    # MSI Authentication Token
+    resource = '20e940b3-4c77-4b0b-9a53-9e16a1b010a7'
+    client_id = os.environ['CLIENT_ID']
+    url = (
+        f"{METADATA_URL}"
+        f"identity/oauth2/token?api-version=2018-02-01&client_id={client_id}"
+        f"&resource={resource}"
+    )
+    try:
+        auth_token = json.loads(_fetch_metadata(url))
+        if auth_token["token_type"] == "Bearer" and auth_token["access_token"]:
```

Comment:
Nitpick: space before and after conditional for readability.

Reply:
What do  you mean, there is a space already around the `==`  ?

Reply:
The full conditional:

```
# stuff

if something:
    do something

# more stuff
```

Reply:
oh you meant blank lines.. I was confused as it mentions "space", i.e. whitespace. I will fix this as well

Reply:
yep, sorry wasn't clear on that.